### PR TITLE
Add proxy support to RemoteClusterConnection

### LINKED ISSUE #31840: Add proxy mode to RemoteClusterConnection
Today we assume we can use the internal structure of a remote cluster to make additional connections. Yet, this might not work if the remote cluster is inside a cloud-env or behind a proxy. We can't use a seed node and then sniff / discover from there using the cluster state of the remote cluster. 

One way to solve this would be a proxy mode for remote connections that doesn't sniff the remote cluster at all but uses N connections to a proxy using dummy discovery nodes internally. Configuration wise this could look like this:

```YAML
search:
    remote:
        cluster_one: 
            seeds: 127.0.0.1:9300
            proxy_mode: true
            num_connections: 10
```

This would translate to connecting to 10 dummy DiscoveryNodes using the same seed which can then round robin or whatever to connect to nodes. Each of these dummies would use a single channel profile. It's quite a different way of connecting but should work just fine for CCS/CCR purposes.

/cc @tbrooks8 @bleskes 

Relates to #31835

### PR DESCRIPTION
**Pull request by s1monw**
This adds support for connecting to a remote cluster through
a tcp proxy. A remote cluster can configured with an additional
`search.remote.$clustername.proxy` setting. This proxy will be used
to connect to remote nodes for every node connection established.
We still try to sniff the remote clsuter and connect to nodes directly
through the proxy which has to support some kind of routing to these nodes.
Yet, this routing mechanism requires the handshake request to include some
kind of information where to route to which is not yet implemented. The effort
to use the hostname and an optional node attribute for routing is tracked
in #32517

Closes #31840

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/transport/RemoteClusterAware.java [hunk 1]
```diff
@@ -77,25 +97,37 @@ protected RemoteClusterAware(Settings settings) {
         this.clusterNameResolver = new ClusterNameExpressionResolver(settings);
     }
 
-    protected static Map<String, List<Supplier<DiscoveryNode>>> buildRemoteClustersSeeds(Settings settings) {
+    protected static Map<String, Tuple<String, List<Supplier<DiscoveryNode>>>> buildRemoteClustersSeeds(Settings settings) {
```
Comment (tbrooks8) [inline server/src/main/java/org/elasticsearch/transport/RemoteClusterAware.java]: Maybe a doc comment here? With the new work, it is really tough to figure out what is going on in this method.

File: server/src/test/java/org/elasticsearch/transport/RemoteClusterConnectionTests.java [hunk 2]
```diff
@@ -1384,4 +1391,98 @@ public void testLazyResolveTransportAddress() throws Exception {
...
+            }
+        }
+    }
+
+    public static Transport getProxyTransport(ThreadPool threadPool, Map<String, Map<String, DiscoveryNode>> nodeMap) {
+        if (nodeMap.isEmpty()) {
+            throw new IllegalArgumentException("nodeMap must be non-empty");
+        }
+
+        return new StubbableTransport(MockTransportService.newMockTransport(Settings.EMPTY, Version
```
Comment (tbrooks8) [inline server/src/test/java/org/elasticsearch/transport/RemoteClusterConnectionTests.java]: Can we not extend and override `StubbableTransport` like this? Ideally maybe the class should be final. It provides methods to attach lambdas for "stubbing" the behavior (although I think the method will need to be made public). The method is either `addConnectBehavior` or you can add a `setDefaultConnectbehavior`. Similar to `setDefaultSendBehavior`.

**Review by Tim-Brooks — COMMENTED** [review]
Two comments.

File: server/src/test/java/org/elasticsearch/transport/RemoteClusterConnectionTests.java (hunk 2 above)
Comment (tbrooks8) [reply server/src/test/java/org/elasticsearch/transport/RemoteClusterConnectionTests.java]: The theoretical idea here is to try to move away from overriding methods like crazy at the transport level. So if refactorings need to happen, we can (hopefully) just move the stubs to different locations, opposed to dealing with a billion different tightly couple to the `Transport` interface tests.

**Comment by s1monw** [comment]
@tbrooks8 I pushed changes for your feedback

File: server/src/test/java/org/elasticsearch/transport/RemoteClusterConnectionTests.java (hunk 2 above)
Comment (s1monw) [reply server/src/test/java/org/elasticsearch/transport/RemoteClusterConnectionTests.java]: ++ good call. I made the class final.

**Review by Tim-Brooks — APPROVED** [review]
LGTM

**Review by jaymode — APPROVED** [review]
LGTM

**Pull request merged**